api/v1alpha1/ratelimitfilter_types.go

```diff
@@ -0,0 +1,130 @@
+type RateLimitFilter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RateLimitFilterSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitFilterList contains a list of RateLimitFilter resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimitFilter `json:"items"`
+}
+
+// RateLimitFilterSpec defines the desired state of RateLimitFilter
+type RateLimitFilterSpec struct {
+	// Type decides the scope for the RateLimits.
+	Type RateLimitType `json:"type"`
+	// Rules are a list of RateLimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// Valid RateLimitType values are:
+//
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	RateLimitTypeGlobal RateLimitType = "Global"
+)
+
+// RateLimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RateLimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RateLimitMatch `json:"matches,omitempty"`
+	Limit   RateLimitValue   `json:"limit"`
+}
+
+// RateLimitMatch specifies the attributes within the traffic flow that can
+// be matched on.
+type RateLimitMatch struct {
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []HeaderMatch `json:"headers,omitempty"`
+}
+
+// HeaderMatch defines the match attributes within the HTTP Headers of the request.
+type HeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name of the HTTP header.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	Name string `json:"name"`
+
+	// Value within the HTTP header. Due to the
+	// case-insensitivity of header names, "foo" and "Foo" are considered equivalent.
+	// Do not set this field when Type="Any", implying matching on any/all unique values within the header.
+	// +optional
+	Value *string `json:"value,omitempty"`
+}
+
+// HeaderMatchType specifies the semantics of how HTTP header values should be
+// compared. Valid HeaderMatchType values are:
+//
+// * "Exact": Use this type to match the exact value of the Value field against the value of the specified HTTP Header.
+// * "RegularExpression": Use this type to match a regular expression against the value of the specified HTTP Header.
+// * "Any": Use this type to match any and all possible unique values encountered in the specified HTTP Header.
```

Comment:
possible values here could be "Any", "All", "PerUniqueValue", went with "Any" for now. this type helps the user simplify
writing down intent, instead of specifying all possible unique values in the config (for attributes such as user-id), this match
type could be used

Reply:
Maybe just a little bit more about the fact that _each unique value_ will have a separate rate limit?

Something like:
```suggestion
// * "Any": Use this type to match any and all possible unique values encountered in the specified HTTP Header.
//   Note that each unique value will receive its own rate limit bucket.
```

Reply:
ah thanks, adding this !

Reply:
I really like the suggestion mentioned [here](https://github.com/envoyproxy/gateway/pull/706#pullrequestreview-1214295911) of using `Distinct` as a keyword to describe this intent, thanks @kflynn !